internal/warehouse/githubActions.go

```diff
@@ -0,0 +1,394 @@
+	var err error
+	var resp *github.Response
+	var workflowRunsPage *github.WorkflowRuns
+	opt := &github.ListWorkflowRunsOptions{
+		ListOptions: github.ListOptions{PerPage: 30},
+	}
+
+	for _, workflow := range workflowsPage {
+		w.logger.Info().Msgf("getting workflow runs and workflow jobs for workflow %s", *workflow.Name)
+		// we get a page of 30 workflow runs until next page is 0
+		for {
+
+			if workflowRunsPage, resp, err = w.githubClient.Actions.ListWorkflowRunsByID(ctx, owner, repo, *workflow.ID, opt); err != nil {
+				return err
+			}
+
+			if *workflowRunsPage.TotalCount > 0 && len(workflowRunsPage.WorkflowRuns) > 0 {
+
+				if err := w.handleWorkflowRunsUpsert(ctx, workflowRunsPage.WorkflowRuns, repoID); err != nil {
+					return err
+				}
+
+				if err := w.handleWorkflowRunsJobs(ctx, owner, repo, repoID, workflowRunsPage.WorkflowRuns); err != nil {
+					return err
+				}
+			}
+
+			w.restRatelimitHandler(ctx, resp)
+
+			if resp.NextPage == 0 {
+				break
+			}
+
+			opt.Page = resp.NextPage
+
+		}
+
+		opt.Page = 0
+
+	}
+
+	return nil
+}
+
+func (w *warehouse) handleWorkflowRunsJobs(ctx context.Context, owner, repo string, repoID uuid.UUID, workflowRunsPage []*github.WorkflowRun) error {
+	var err error
+	var resp *github.Response
+	var workflowRunJobsPage *github.Jobs
+
+	opt := &github.ListWorkflowJobsOptions{
+		ListOptions: github.ListOptions{PerPage: 30},
+	}
+
+	for _, workflowRun := range workflowRunsPage {
+
+		// we get a page of 30 workflow run jobs until next page is 0
+		for {
+			if workflowRunJobsPage, resp, err = w.githubClient.Actions.ListWorkflowJobs(ctx, owner, repo, *workflowRun.ID, opt); err != nil {
+				return err
+			}
+
+			if *workflowRunJobsPage.TotalCount > 0 && len(workflowRunJobsPage.Jobs) > 0 {
+				if err := w.handleWorkflowJobLogs(ctx, owner, repo, repoID, workflowRunJobsPage.Jobs); err != nil {
+					return err
+				}
+			}
+
+			w.restRatelimitHandler(ctx, resp)
+
+			if resp.NextPage == 0 {
+				break
+			}
+
+			opt.Page = resp.NextPage
+
+		}
+
+		// we reset the page for the next run
+		opt.Page = 0
+	}
+
+	return nil
+}
+
+func (w *warehouse) handleWorkflowJobLogs(ctx context.Context, owner, repo string, repoID uuid.UUID, workflowRunJobsPage []*github.WorkflowJob) error {
+	var err error
+	var resp *github.Response
+	var log string
+	var workflowJobLog *url.URL
+
+	// we iterate over the workflowrunJobs page to get each log
+	for i, workflowJob := range workflowRunJobsPage {
+
+		if workflowJobLog, resp, err = w.githubClient.Actions.GetWorkflowJobLogs(ctx, owner, repo, *workflowJob.ID, true); err != nil {
+			return err
+		}
+
+		if len(workflowJobLog.String()) > 0 {
+			log, err = w.parseJobLogs(workflowJobLog, strconv.Itoa(int(*workflowJob.ID)))
+			if err != nil {
+				return err
+			}
+		}
+
+		if err := os.RemoveAll(strconv.Itoa(i)); err != nil {
+			w.logger.Err(err)
+			return err
+		}
```

Comment:
So I think the removal of the file should be handled closer to where we create/download the file. In `parseJobLogs`, if we succeed in downloading the file, but then encounter an error reading it (when we call `ReadFile`), then we won't reach this part of the code to cleanup/remove that file, and it will get left behind on disk.

A pattern that could be used is returning a "cleanup" `func` from the `parseJobLogs` which can be called in a `defer` (so that it's always called when the function exits, either with an error or not). This approach is taken [here](https://github.com/mergestat/mergestat/blob/main/internal/syncer/syncer.go#L215) when we clone repos and then clean them up after a sync run (making sure to clean it up even if an error occurs)

My other request here would be that we use a temporary directory to download the file to, again configured similarly to how we do git clones: https://github.com/mergestat/mergestat/blob/main/internal/syncer/syncer.go#L216

as that will allow us to control where these temporary files go on disk (which should be configurable in our deployment somehow)

Reply:
For  this iteration we will keep the grap library due to small  issues of our current approach vs the standard  library approach , however we are using the temp pattern now